CRM/Contribute/BAO/Contribution.php

```diff
@@ -5005,11 +5008,17 @@ public static function checkLineItems(&$params) {
         $lineItemAmount += $item['line_total'];
       }
     }
+
     if (!isset($totalAmount)) {
       $params['total_amount'] = $lineItemAmount;
     }
-    elseif ($totalAmount != $lineItemAmount) {
-      throw new API_Exception("Line item total doesn't match with total amount.");
+    else {
+      $difference = abs($totalAmount - $lineItemAmount);
+      $precision = 0.001;
```

Comment:
@mlutfy this is hardcoding in a precision which should really depend on the number of digits in the currency. Could we create a setting for this and default it to the appropriate precision for two digit after the decimal.

@MiyaNoctem can you explain to me why you have chosen three digits of accuracy rather than two? We should only be storing two digits in the database, even if during some calculations interim values use one extra digit.

Reply:
@joemurray Could you read the other review comments in this PR please?  It needs to be updated to use the *existing* functions for calculating money and the *existing* functions for getting precision.

When performing a calculation you need to work to one point of accuracy greater than the intended result or you will get rounding errors. That is why @MiyaNoctem is using three digits of accuracy.

Reply:
@mattwire I'm commenting on the current code in the PR. I don't have a view on your comment about helper function. I don't believe it is good practice to be leaving interim calculation values in a variable across function boundaries. These should be money values that have two digits of precision after the decimal point for the default currency. For addition operations of line item totals to get the invoice total it is *not* appropriate to be leaving an additional digit of precision because the total will possibly not equal the sum of the actual money amounts of the line items when there are a few dozen or more line items.

Reply:
@mattwire, I tried adjusting the code to use the method you suggested, but it actually doesn't work still, even taking into account the precision, a very minuscule difference is still found. This is the code I tried:

```php
      $adjustedDifference = CRM_Utils_Money::subtractCurrencies($totalAmount, $lineItemAmount, $params['currency']);

      if ($adjustedDifference != 0) {
        throw new API_Exception(self::LINE_ITEM_DIFFERRING_TOTAL_EXCEPTON_MSG);
      }
```

When I tried it with the test, `$adjustedDifference = 4.54E-15`, still differento to 0. I also tried doing the subtraction first, and then comparing the difference with 0 by calling `CRM_Utils_Money::subtractCurrencies()`, but that also failed, although with a smaller difference (3.55E-15).

So I created an equals method on the Money class, that follows suit as to how subtract method works. Although I understand @JoeMurray's point...